### PR TITLE
Add WebSocket endpoint for adb proxy

### PR DESCRIPTION
**Pull request by k311093**
Add url /devices/dev_id/adb as WebSocket endpoint that proxies WebSocket payloads to the tcp port for the adb of device for dev_id.

### CODE REVIEW

File: frontend/src/liboperator/operator/operator.go [hunk 1]
```diff
@@ -480,6 +484,111 @@ func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
...
+	adbPort := 0
+	if adb_port, ok := devInfo["adb_port"]; ok {
+		adbPort = int(adb_port.(float64))
+	} else {
+		for i, id := range pool.DeviceIds() {
+			if id == devId {
+				adbPort = 6520 + i
+				break
+			}
+		}
```
Comment (0405ysj) [inline frontend/src/liboperator/operator/operator.go:512]: I think this fallback mechanism is not reliable. Especially when the base instance number is not 1. Shouldn't we make backporting changes for previous CF releases instead?

File: frontend/src/liboperator/operator/operator.go [hunk 2]
```diff
@@ -480,6 +484,111 @@ func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
...
+		log.Print("Error while connect to ADB: ", err)
+		return
+	}
+	defer tcpConn.Close()
+
+	// Redirect WebSocket to ADB tcp socket
+	redirectWsToTcp(wsConn, tcpConn)
+}
+
+func redirectWsToTcp(wsConn *websocket.Conn, tcpConn net.Conn) {
```
Comment (0405ysj) [inline frontend/src/liboperator/operator/operator.go:539]: Couldn't bi-directional data transmission be happened with `io.copy`? Just out of curiosity, I think it would make code much simpler if it's available.

**Review by 0405ysj — CHANGES_REQUESTED** [review]

File: frontend/src/liboperator/operator/operator.go (hunk 2 above)
Comment (k311093) [reply frontend/src/liboperator/operator/operator.go]: Changed to use io.Copy.

File: frontend/src/liboperator/operator/operator.go (hunk 1 above)
Comment (k311093) [reply frontend/src/liboperator/operator/operator.go:513]: @ser-io @jemoreira I have some question about this.

Does it okay to move this logic to orchestrator? or it needs to be in operator. Since operator does not have adb connection info for the device I added adb port info to device registration info, but it might be a problem when device is launched with old host packages.

I added the fallback logic, but there are some situations this might not work.

But if this logic is in the orchestrator, it might be possible to get adb_serial info from orchestrator.

Do you have any suggestions for that?

File: frontend/src/liboperator/operator/operator.go [hunk 3]
```diff
@@ -480,6 +485,100 @@ func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
...
+type wsIoWrapper struct {
+	wsConn *websocket.Conn
+	pos int
+	buf []byte
+}
+
+var _ io.ReadWriteCloser = (*wsIoWrapper)(nil)
+
+func (w *wsIoWrapper) Read(buf []byte) (int, error) {
+	if w.buf == nil || w.pos == len(w.buf) {
```
Comment (0405ysj) [inline frontend/src/liboperator/operator/operator.go:557]: nit: `w.pos >= len(w.buf)`?

File: frontend/src/liboperator/operator/operator.go [hunk 4]
```diff
@@ -480,6 +485,100 @@ func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
...
+
+func (w *wsIoWrapper) Write(buf []byte) (int, error) {
+	err := w.wsConn.WriteMessage(websocket.BinaryMessage, buf)
+	if err != nil {
+		return 0, err
+	}
+	return len(buf), nil
+}
+
+func (w *wsIoWrapper) Close() error {
```
Comment (0405ysj) [inline frontend/src/liboperator/operator/operator.go:578]: When is it used?

File: frontend/src/liboperator/operator/operator.go [hunk 5]
```diff
@@ -480,6 +485,100 @@ func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
...
+	wsConn *websocket.Conn
+	pos int
+	buf []byte
+}
+
+var _ io.ReadWriteCloser = (*wsIoWrapper)(nil)
+
+func (w *wsIoWrapper) Read(buf []byte) (int, error) {
+	if w.buf == nil || w.pos == len(w.buf) {
+		_, buf, err := w.wsConn.ReadMessage()
```
Comment (0405ysj) [inline frontend/src/liboperator/operator/operator.go:558]: As we talked, the variable name `buf` seems duplicated.

File: frontend/src/liboperator/operator/operator.go (hunk 3 above)
Comment (k311093) [reply frontend/src/liboperator/operator/operator.go]: Changed to use the expression.

File: frontend/src/liboperator/operator/operator.go (hunk 5 above)
Comment (k311093) [reply frontend/src/liboperator/operator/operator.go]: Changed buf name that is duplicated.

File: frontend/src/liboperator/operator/operator.go (hunk 4 above)
Comment (k311093) [reply frontend/src/liboperator/operator/operator.go]: Close() is not used, removed the function and changed implementing interface from io.ReadWriteCloser to io.ReadWriter.

File: frontend/src/liboperator/operator/operator.go [hunk 6]
```diff
@@ -480,6 +485,104 @@ func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
...
+	// Redirect WebSocket to ADB tcp socket
+	wsWrapper := &wsIoWrapper{
+		wsConn: wsConn,
+		pos: 0,
+		buf: nil,
+	}
+
+	go func() {
+		io.Copy(wsWrapper, tcpConn)
+		wsWrapper.Close()
```
Comment (0405ysj) [inline frontend/src/liboperator/operator/operator.go:545]: As we talked before, please ensure both connection being closed when it ends.

File: frontend/src/liboperator/operator/operator.go [hunk 7]
```diff
@@ -480,6 +485,104 @@ func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
...
+	}()
+	io.Copy(tcpConn, wsWrapper)
+	tcpConn.Close()
```
Comment (0405ysj) [inline frontend/src/liboperator/operator/operator.go:548]: ditto.

File: frontend/src/liboperator/operator/operator.go [hunk 8]
```diff
@@ -480,6 +485,104 @@ func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
...
+}
+
+// Wrapper for implementing io.ReadWriteCloser of websocket.Conn
+type wsIoWrapper struct {
+	wsConn *websocket.Conn
+	pos int
+	buf []byte
```
Comment (0405ysj) [inline frontend/src/liboperator/operator/operator.go:555]: nit: Could you leave a short comment that those fields are for `Read` function only?

File: frontend/src/liboperator/operator/operator.go (hunk 1 above)
Comment (0405ysj) [reply frontend/src/liboperator/operator/operator.go:513]: AFAIK, we would unite Operator and Host Orchestrator into one. (Correct me if I'm wrong.)

Then, it's better to be in Host Orchestrator, since it holds information about adb_serial.
TBH, we don't need ADB port but ADB serial instead, isn't it?

**Review by 0405ysj — CHANGES_REQUESTED** [review]

File: frontend/src/liboperator/operator/operator.go (hunk 6 above)
Comment (k311093) [reply frontend/src/liboperator/operator/operator.go:545]: This code had already patched for connection closing.

File: frontend/src/liboperator/operator/operator.go (hunk 7 above)
Comment (k311093) [reply frontend/src/liboperator/operator/operator.go:548]: This code had also patched for connection closing.

File: frontend/src/liboperator/operator/operator.go (hunk 8 above)
Comment (k311093) [reply frontend/src/liboperator/operator/operator.go]: Added comment for the field that is used for Read

File: frontend/src/liboperator/operator/operator.go (hunk 1 above)
Comment (k311093) [reply frontend/src/liboperator/operator/operator.go:513]: If an instance is started by `cvd start` it would be fine. adb_serial for host orchestrator is got by `cvd fleet` which shows instances created by `cvd start` (of course created by host orchestrator).

If supporting ADB connection proxy from host orchestrator it might need to rename `/devices/{deviceId}/adb` endpoint to `cvds/{group}/{name}/adb`.

Or another option might be just return an error `Device is not supported for ADB` when deviceInfo does not have ADB port for operator, and backport ADB port registration change(http://aosp/3154596) for the device when it is needed.

@ser-io @jemoreira That do you think?

Comment (k311093) [reply frontend/src/liboperator/operator/operator.go]: Decided not to use fallback logic but just return errors when adb_port is not registered.

**Review by 0405ysj — APPROVED** [review]

**Comment by ser-io** [comment]
Team remember to use `git commit --amend` and `git push -f` when addressing comments, it makes the history of the repository cleaner and the PRs easier to review. See [PR#307](https://github.com/google/cloud-android-orchestration/pull/307) (having 50 threads and only 1 commit) as a reference. Thanks!

File: frontend/src/liboperator/operator/operator.go [hunk 9]
```diff
@@ -533,60 +534,49 @@ func adbProxy(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
...
+}
 
-			wsData := make([]byte, len(buf))
-			copy(wsData, buf)
+var _ io.ReadWriteCloser = (*wsIoWrapper)(nil)
```
Comment (ser-io) [inline frontend/src/liboperator/operator/operator.go:548]: why is this needed?

File: frontend/src/liboperator/operator/operator.go [hunk 10]
```diff
@@ -533,60 +534,49 @@ func adbProxy(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
...
+	if w.buf == nil || w.pos == len(w.buf) {
+		_, buf, err := w.wsConn.ReadMessage()
+		if err != nil {
+			return 0, err
 		}
-	}(wsConn, wsChan)
+		w.buf = buf
+		w.pos = 0
+	}
+	nRead := copy(buf[:], w.buf[w.pos:])
```
Comment (ser-io) [inline frontend/src/liboperator/operator/operator.go:559]: we can use `buf` directly here as it's already an slice.

File: frontend/src/liboperator/operator/operator.go [hunk 11]
```diff
@@ -533,60 +534,49 @@ func adbProxy(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
...
+	pos int
+	buf []byte
+}
 
-			wsData := make([]byte, len(buf))
-			copy(wsData, buf)
+var _ io.ReadWriteCloser = (*wsIoWrapper)(nil)
 
-			wsChan <- wsData
+func (w *wsIoWrapper) Read(buf []byte) (int, error) {
```
Comment (ser-io) [inline frontend/src/liboperator/operator/operator.go:550]: nit: use `p` argument name rather than `buf` to avoid confusion with `w.buf` using the name in the original signature: `Read(p []byte) (n int, err error)` https://pkg.go.dev/io#Reader

**Review by ser-io — COMMENTED** [review]
We should use `git commit --amend` and `git push -f `when addressing comments and avoid new commits, it makes the history of the repository cleaner and the PRs easier to review. See https://github.com/google/cloud-android-orchestration/pull/307 (having 50 threads and only 1 commit) as a reference.

File: frontend/src/liboperator/operator/operator.go (hunk 9 above)
Comment (k311093) [reply frontend/src/liboperator/operator/operator.go:548]: This code is for confirming that `wsIoWrapper` implements `io.ReadWriteCloser` interface

File: frontend/src/liboperator/operator/operator.go (hunk 10 above)
Comment (k311093) [reply frontend/src/liboperator/operator/operator.go:559]: I'll upload separated PR for this.

File: frontend/src/liboperator/operator/operator.go (hunk 11 above)
Comment (k311093) [reply frontend/src/liboperator/operator/operator.go:550]: I'll upload separated PR for this.

**Comment by k311093** [comment]
> We should use `git commit --amend` and `git push -f `when addressing comments and avoid new commits, it makes the history of the repository cleaner and the PRs easier to review. See [google/cloud-android-orchestration#307](https://github.com/google/cloud-android-orchestration/pull/307) (having 50 threads and only 1 commit) as a reference.

Acknowledged, thanks!

File: frontend/src/liboperator/operator/operator.go (hunk 9 above)
Comment (ser-io) [reply frontend/src/liboperator/operator/operator.go:548]: I see, from https://go.dev/doc/effective_go#interfaces_and_types this is needed when interface checks happens at runtime which is not the case here. 

```
The appearance of the blank identifier in this construct indicates that the declaration exists only for the type checking, not to create a variable. Don't do this for every type that satisfies an interface, though. By convention, such declarations are only used when there are no static conversions already present in the code, which is a rare event.
```

File: frontend/src/liboperator/operator/operator.go [hunk 12]
```diff
@@ -480,6 +485,94 @@ func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
...
+		wsConn: wsConn,
+		pos:    0,
+		buf:    nil,
+	}
+	defer wsWrapper.Close()
+
+	// Redirect WebSocket to ADB tcp socket
+	go func() {
+		io.Copy(wsWrapper, tcpConn)
+		wsWrapper.Close()
```
Comment (jemoreira) [inline frontend/src/liboperator/operator/operator.go:534]: Should this be `tcpConn.Close()`? The `wsWrapper` is already closed on line 529.

**Comment by ikicha** [comment]
If we have websocket endpoints for adb, can we simplify cvdr impl for adb connect? WDYT @jemoreira @0405ysj

**Comment by jemoreira** [comment]
> If we have websocket endpoints for adb, can we simplify cvdr impl for adb connect? WDYT @jemoreira @0405ysj

Websocket doesn't work in AppEngine, so we need both implementations in cvdr unfortunately.

**Comment by 0405ysj** [comment]
> If we have websocket endpoints for adb, can we simplify cvdr impl for adb connect?

What I heard from Jaeman was, there are 2 types of AppEngine environments; standard vs flexible.
Standard is what we're currently using, but it doesn't support websocket.
Flexible supports websocket, but this type is not available for the internal uses.

AFAIK Jaeman is currently investigating around AppEngine deeply.

File: frontend/src/liboperator/operator/operator.go (hunk 12 above)
Comment (k311093) [reply frontend/src/liboperator/operator/operator.go:534]: When I tested it, deferred wsWrapper.Close() was not called with spawned go routine, so it need to close explicitly io.Copy() call at line 536 to be exited when tcpConn is closed.
